File: python/src/ttio/uv_vis_spectrum.py
"""``UVVisSpectrum`` — 1-D UV-visible absorption spectrum subclass."""
from __future__ import annotations

from dataclasses import dataclass

from .signal_array import SignalArray
from .spectrum import Spectrum


@dataclass(slots=True)
class UVVisSpectrum(Spectrum):
    """A 1-D UV-Vis spectrum with ``wavelength`` (nm) and ``absorbance`` channels.

    Unlike Raman / IR spectra, UV-Vis observations are indexed by
    wavelength in nanometres rather than wavenumber.

    Parameters
    ----------
    path_length_cm : float, default 0.0
        Optical path length of the cuvette in centimetres.
    solvent : str, default ``""``
        Free-form solvent description (e.g. ``"methanol"``).
    *plus all base :class:`Spectrum` parameters*

    Notes
    -----
    API status: Stable (v0.11.1).

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIOUVVisSpectrum`` · Java:
    ``global.thalion.ttio.UVVisSpectrum``.
    """

    WAVELENGTH = "wavelength"
    ABSORBANCE = "absorbance"

    path_length_cm: float = 0.0
    solvent: str = ""

    @property
    def wavelength_array(self) -> SignalArray:
        """Return the ``"wavelength"`` :class:`SignalArray`."""
        return self.signal_array(self.WAVELENGTH)

    @property
    def absorbance_array(self) -> SignalArray:
        """Return the ``"absorbance"`` :class:`SignalArray`."""
        return self.signal_array(self.ABSORBANCE)

File: java/src/main/java/global/thalion/ttio/UVVisSpectrum.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio;

import java.util.Map;

/**
 * 1-D UV-visible absorption spectrum: wavelength (nm) and absorbance
 * channels plus path-length and solvent metadata.
 *
 * <p><b>API status:</b> Stable (v0.11.1).</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOUVVisSpectrum}, Python
 * {@code ttio.uv_vis_spectrum.UVVisSpectrum}.</p>
 *
 *
 */
public class UVVisSpectrum extends Spectrum {
    private final double pathLengthCm;
    private final String solvent;

    public UVVisSpectrum(double[] wavelengthValues, double[] absorbanceValues,
                         int indexPosition, double scanTimeSeconds,
                         double pathLengthCm, String solvent) {
        super(Map.of(
            "wavelength", SignalArray.ofDoubles(wavelengthValues),
            "absorbance", SignalArray.ofDoubles(absorbanceValues)
        ), indexPosition, scanTimeSeconds);
        this.pathLengthCm = pathLengthCm;
        this.solvent = solvent == null ? "" : solvent;
    }

    /** Returns the {@code "wavelength"} {@link SignalArray}. */
    public SignalArray wavelengthArray() { return signalArray("wavelength"); }

    /** Returns the {@code "absorbance"} {@link SignalArray}. */
    public SignalArray absorbanceArray() { return signalArray("absorbance"); }

    /** Convenience: wavelength values as a primitive array. */
    public double[] wavelengthValues() { return wavelengthArray().asDoubles(); }

    /** Convenience: absorbance values as a primitive array. */
    public double[] absorbanceValues() { return absorbanceArray().asDoubles(); }

    /** Cuvette path length in centimetres. */
    public double pathLengthCm() { return pathLengthCm; }

    /** Solvent description (free-form). */
    public String solvent() { return solvent; }
}
